internal/orchestrator/reorg_handler.go

```diff
@@ -0,0 +1,187 @@
+	if !config.Cfg.ReorgHandler.ForceFromBlock {
+		storedFromBlock, err := storage.OrchestratorStorage.GetLastReorgCheckedBlockNumber(chainId)
+		if err != nil {
+			log.Debug().Err(err).Msgf("Error getting last reorg checked block number, using configured: %s", bn)
+			return bn
+		}
+		if storedFromBlock.Sign() <= 0 {
+			log.Debug().Msgf("Last reorg checked block number not found, using configured: %s", bn)
+			return bn
+		}
+		log.Debug().Msgf("Last reorg checked block number found, using: %s", storedFromBlock)
+		return storedFromBlock
+	}
+	log.Debug().Msgf("Force from block reorg check flag set, using configured: %s", bn)
+	return bn
+}
+
+func (rh *ReorgHandler) Start() {
+	interval := time.Duration(rh.triggerInterval) * time.Millisecond
+	ticker := time.NewTicker(interval)
+
+	log.Debug().Msgf("Reorg handler running")
+	go func() {
+		for range ticker.C {
+			lookbackFrom := new(big.Int).Add(rh.lastCheckedBlock, big.NewInt(int64(rh.blocksPerScan)))
+			blockHeaders, err := rh.storage.MainStorage.LookbackBlockHeaders(rh.rpc.ChainID, rh.blocksPerScan, lookbackFrom)
+			if err != nil {
+				log.Error().Err(err).Msg("Error getting recent block headers")
+				continue
+			}
+			reorgEndIndex := findReorgEndIndex(blockHeaders)
+			if reorgEndIndex == -1 {
+				rh.lastCheckedBlock = lookbackFrom
+				rh.storage.OrchestratorStorage.SetLastReorgCheckedBlockNumber(rh.rpc.ChainID, lookbackFrom)
+				metrics.ReorgHandlerLastCheckedBlock.Set(float64(lookbackFrom.Int64()))
+				continue
+			}
+			metrics.ReorgCounter.Inc()
+			forkPoint, err := rh.findForkPoint(blockHeaders[reorgEndIndex:])
+			if err != nil {
+				log.Error().Err(err).Msg("Error while finding fork point")
+				continue
+			}
+			err = rh.handleReorg(forkPoint, lookbackFrom)
+			if err != nil {
+				log.Error().Err(err).Msg("Error while handling reorg")
+				continue
+			}
+			rh.lastCheckedBlock = lookbackFrom
+			rh.storage.OrchestratorStorage.SetLastReorgCheckedBlockNumber(rh.rpc.ChainID, lookbackFrom)
+			metrics.ReorgHandlerLastCheckedBlock.Set(float64(lookbackFrom.Int64()))
+		}
+	}()
+
+	// Keep the program running (otherwise it will exit)
+	select {}
+}
+
+func findReorgEndIndex(reversedBlockHeaders []common.BlockHeader) (index int) {
+	for i := 0; i < len(reversedBlockHeaders)-1; i++ {
+		currentBlock := reversedBlockHeaders[i]
+		previousBlock := reversedBlockHeaders[i+1]
+
+		if currentBlock.ParentHash != previousBlock.Hash {
+			log.Info().
+				Str("currentBlockNumber", currentBlock.Number.String()).
+				Str("currentBlockHash", currentBlock.Hash).
+				Str("currentBlockParentHash", currentBlock.ParentHash).
+				Str("previousBlockNumber", previousBlock.Number.String()).
+				Str("previousBlockHash", previousBlock.Hash).
+				Msg("Reorg detected: parent hash mismatch")
+			return i
+		}
+	}
+	return -1
+}
+
+func (rh *ReorgHandler) findForkPoint(reversedBlockHeaders []common.BlockHeader) (forkPoint *big.Int, err error) {
```

Comment:
This function is hard to follow with all the loops + recursion. It will help splitting it up to smaller functions that show the steps taken to find the fork point

Reply:
Separated the fetching of new blocks